server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -57,12 +60,59 @@ public CoordinatorStats run(DruidCoordinator coordinator, DruidCoordinatorRuntim
 
     final Map<String, Integer> loadStatus = Maps.newHashMap();
 
+    final Map<String, Integer> tieredReplicants = getTieredReplicants();
+    for (final String tier : tieredReplicants.keySet()) {
+      stats.addToTieredStat(ASSIGNED_COUNT, tier, 0);
+    }
+
+    final BalancerStrategy strategy = params.getBalancerStrategy();
+
+    final int maxSegmentsInNodeLoadingQueue = params.getCoordinatorDynamicConfig()
+                                                    .getMaxSegmentsInNodeLoadingQueue();
+
+    final Predicate<ServerHolder> serverHolderPredicate;
+    if (maxSegmentsInNodeLoadingQueue > 0) {
+      serverHolderPredicate = s -> (s != null && s.getNumberOfSegmentsInQueue() < maxSegmentsInNodeLoadingQueue);
+    } else {
+      serverHolderPredicate = Objects::nonNull;
+    }
+
     int totalReplicantsInCluster = params.getSegmentReplicantLookup().getTotalReplicants(segment.getIdentifier());
-    for (Map.Entry<String, Integer> entry : getTieredReplicants().entrySet()) {
+
+    final Optional<ServerHolder> primaryHolderToLoad;
+    if (totalReplicantsInCluster <= 0) {
+      log.trace("No replicants for %s", segment.getIdentifier());
+      primaryHolderToLoad = getPrimaryHolder(
+          params.getDruidCluster(),
+          tieredReplicants,
+          strategy,
+          segment,
+          serverHolderPredicate
+      );
+
+      if (primaryHolderToLoad.isPresent()) {
```

Comment:
this feels strange to not use a functional workflow here, but it is not clear a functional flow would be easier to read.

Reply:
On the contrary, I removed use of `Optional` and mapping, because IMO it only adds obscurity. "Too functional" for Java